### ISSUE TITLE
Can we switch to hosting the DUB binaries on GitHub?

### ISSUE THREAD

**Issue by wilzbach** [dlang/dub#1290]
tl;dr: code.dlang.org is often down [1, 2]. While we now have code-mirror{2,3}.dlang.io for the repos, we don't have a mirror for the DUB binaries. [While it would be possible to avoid downloading DUB on Travis](https://github.com/dlang/installer/pull/218) as it's bundled with dmd, it doesn't seem that we will do this switch soon. Hence, I propose to use the GitHub release feature as it's very convenient (either drag & drop the binaries or use their API), free, very reliable and many tools for it exist in the wild. Even the ldc developers [use GitHub releases](https://github.com/ldc-developers/ldc-scripts/blob/master/ldc2-packaging/4-push-to-github.sh) for more than four years now.

![image](https://user-images.githubusercontent.com/4370550/33328677-e3807c3e-d45a-11e7-9b02-ed411be9442b.png)

I'm more then happy to help uploading, building or mirroring the binaries. @s-ludwig can I start by copying the binaries from http://code.dlang.org/download to their respective GitHub releases?

CC @s-ludwig @JackStouffer

[1] http://forum.dlang.org/post/ptjbjluatugaozjxxvta@forum.dlang.org
[2] http://forum.dlang.org/post/bbfgoochvnublwpyctob@forum.dlang.org

**Comment by JackStouffer** [comment]
Why aren't we just using the D Foundation to pay for an S3 server for the binaries to be hosted on?

**Comment by wilzbach** [comment]
> Why aren't we just using the D Foundation to pay for an S3 server for the binaries to be hosted on?

GitHub uses S3 buckets behind the hood and it's free & easy to use, so why shouldn't we use it?

**Comment by JackStouffer** [comment]
Because AFAIK it can't be automatic.

Maybe I should refocus the question: why isn't all of our infrastructure for code.dlang.org, including the web server, on something like AWS when it's part of our main value proposition with D?

I don't think it's acceptable that our suggested module system has less than 99.9% downtime.

I think we didn't do it in the past because of costs (and believe me, I understand AWS can get expensive). But we have the foundation now.

**Comment by wilzbach** [comment]
> Because AFAIK it can't be automatic.

Huh? There exist plenty of tools - one can even directly use curl:
-> https://github.com/ldc-developers/ldc-scripts/blob/master/ldc2-packaging/4-push-to-github.sh

https://github.com/aktau/github-release

> Maybe I should refocus the question: why isn't all of our infrastructure for code.dlang.org, including the web server, on something like AWS when it's part of our main value proposition with D?

IIRC the connectivity issues are more related to @s-ludwig running Vibe.d without an intermediate layer like a load balancer or caching proxy and thus often running into "new" Vibe.d bugs.
Anyhow you question is quite different and maybe we should continue the discussion on Slack?

**Comment by s-ludwig** [comment]
It would be simple enough to add that to the deployment script, we'd just have to find a replacement for the "LATEST" endpoint that is hosted outside of code.dlang.org. This would definitely be a good idea.

The downtime that happens from time to time currently appears to be related to some kind of memory leak that sneaked in at some point. Unfortunately that makes the server unresponsible, so that it's not a matter of immediate recovery. I'll have too look into watching the memory use manually to shut it down before it reaches ciritical levels, until this is fixed at least.

**Comment by wilzbach** [comment]
> It would be simple enough to add that to the deployment script, 

Is your deployment script available somewhere?
Cool! Can I take this an OK from your side to copy the existing binaries over to GitHub releases at dlang/dub? :)

> we'd just have to find a replacement for the "LATEST" endpoint that is hosted outside of code.dlang.org. This would definitely be a good idea.

Ideas:
- We could use inspiration from the LDC developers again: they simply use GitHub pages: https://github.com/ldc-developers/ldc-developers.github.io
- There's an API to get the latest release (https://developer.github.com/v3/repos/releases/#get-the-latest-release), so using any kind of automation should be simple

**Comment by s-ludwig** [comment]
The script is just on my CI server, but I'll just have to insert the proper curl command. Putting the existing binaries up on GitHub should be absolutely fine, too!

Unfortunately we'd need the latest stable version, so the simple API based route is out I think. Using gh-pages should be fine, the question is where to best automate the upload of that file. Martin's release script would be the obvious choice, but there may be a delay between adding the tag and uploading a new release, so that would be rather error prone unless the binaries are also build by the same script invocation (which kind of makes sense anyway).

**Comment by wilzbach** [comment]
> Putting the existing binaries up on GitHub should be absolutely fine, too!

Okay done.
I also mirrored all files into an S3 bucket as an additional backup / mirror.
For example:

https://s3-eu-west-1.amazonaws.com/code.dlang.org/files/dub-1.6.0-linux-x86_64.tar.gz

I will look into updating the install script, s.t. the GitHub releases are used as a fallback.

**Comment by wilzbach** [comment]
> I will look into updating the install script, s.t. the GitHub releases are used as a fallback.

https://github.com/dlang/installer/pull/273/files

**Comment by MartinNowak** [comment]
We could figure out how to integrate the loose dub binaries with our downloads.dlang.org S3, though GH releases are a good place for those artifacts for now.
Indeed integrating this fully with the dlang release seems like the way to go in the long run.
Not sure what the long-term perspective for separate dub downloads is though?

**Comment by wilzbach** [comment]
> Unfortunately we'd need the latest stable version, so the simple API based route is out I think. Using gh-pages should be fine,

I just started the `gh-pages` branch:

https://github.com/dlang/dub/tree/gh-pages
https://dlang.github.io/dub/LATEST

It's at 1.6.0 atm, because I didn't see release binaries for 1.7.0

> Not sure what the long-term perspective for separate dub downloads is though?

I think the only reason to provide them at the moment is that GDC doesn't ship with a bundled DUB (are they planning to do so with the next big release? CC @ibuclaw).
Of course, there might be a use case to fetch the newest DUB binary if your project is locked to using an old compiler, but those people could simply grab the DUB binary from the newest release.

Building the DUB binaries separately is a lot of redundant work, so I would be in favor of dropping them. In fact we could even do this _now_ - the install script will still fetch dub 1.6.0 for the people testing dmd < 2.072 or gdc on their CI.
Though maybe we should build the binaries for 1.7.0 for the last time?

**Comment by wilzbach** [comment]
@s-ludwig any chance you can integrate the upload to GitHub releases to your release script?
It appears people already consider this as the first spot to look for release binaries:

http://forum.dlang.org/post/lbtbciqactvofnpssiaj@forum.dlang.org

**Comment by jacob-carlborg** [comment]
Travis CI has built-in support for uploading to GitHub releases [1]. See [2] for an example.

[1] https://docs.travis-ci.com/user/deployment/releases/
[2] https://github.com/jacob-carlborg/remarkify/blob/master/.travis.yml#L10

**Comment by wilzbach** [comment]
Yes, I have done this for a few repos @dlang-community too. It's not very hard and thus we can easily avoid things like:

> https://github.com/dlang/dub/releases/latest doesn't point to 1.7.2.

(from: https://forum.dlang.org/post/aymaziydrfhapuiurmpe@forum.dlang.org)
For reference, it's not on code.dlang.org either.

-> let's do this -> https://github.com/dlang/dub/pull/1369

**Comment by Geod24** [comment]
This has been done, and `code.dlang.org` now redirects to Github releases.
